cmd/kubeadm/app/cmd/config.go

```diff
@@ -270,15 +272,62 @@ func NewCmdConfigMigrate(out io.Writer) *cobra.Command {
 func NewCmdConfigUpload(out io.Writer, kubeConfigFile *string) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "upload",
-		Short: "Upload configuration about the current state, so that 'kubeadm upgrade' can later know how to configure the upgraded cluster.",
+		Short: "Uploads the kubeadm and kubelet configuration to a ConfigMap. This is a simple shorthand for the upload-config init phase.",
+		Long:  cmdutil.MacroCommandLongDescription,
 		RunE:  cmdutil.SubCmdRunE("upload"),
 	}
 
-	cmd.AddCommand(NewCmdConfigUploadFromFile(out, kubeConfigFile))
+	// Add legacy deprecated sub-command
 	cmd.AddCommand(NewCmdConfigUploadFromFlags(out, kubeConfigFile))
+
+	// We actually need only the init cmd here, the fake kubeadm cmd is used only for generating correct help messages
+	fakeKubeadmCmd := &cobra.Command{Use: "kubeadm"}
+	fakeKubeadmCmd.AddCommand(NewCmdInit(out, nil))
+
+	// Wrap the upload config init sub-phases into sub-commands
+	uploadConfigPhase := initphases.NewUploadConfigPhase()
+	for _, phase := range uploadConfigPhase.Phases {
+		cmd.AddCommand(configUploadPhaseToCommand(fakeKubeadmCmd, &phase, kubeConfigFile))
+	}
+
+	return cmd
+}
+
+// configUploadPhaseToCommand creates a wrapper cobra command for a given upload-config sub-phase
+func configUploadPhaseToCommand(baseCmd *cobra.Command, phase *phaseworkflow.Phase, kubeConfigFile *string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:                phase.Name,
+		Short:              phase.Short,
+		Long:               phase.Long,
+		Example:            phase.Example,
+		Aliases:            phase.Aliases,
+		DisableFlagParsing: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			runConfigUploadPhase(baseCmd, cmd.Name(), args, *kubeConfigFile)
+		},
+	}
+
+	// setup the extra "from-file" alias to the "kubeadm" phase command
+	if phase.Name == "kubeadm" {
+		cmd.Aliases = append([]string{}, cmd.Aliases...)
```

Comment:
isn't it NOOP?

Reply:
No, this ensures that the append on the following line won't append to `phase.Aliases`.
This is one of the typical Go lang gotchas that one needs to guard against in order to avoid hard to find bugs.

Reply:
Got it. Thanks!